Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -0,0 +1,1011 @@
+           const std::vector<int> &rlabel_attachments) {
+    // some fragments can be add multiple times if they are cyclic
+    for (size_t i=0; i<mols.size(); ++i) {
+      if (newMol.get() == mols[i].get())
+        return;
+    }
+
+    std::copy( rlabel_attachments.begin(), rlabel_attachments.end(),
+               std::inserter( attachments, attachments.end() ) );
+
+    mols.push_back(newMol);
+    std::string smi = MolToSmiles(*newMol, true);
+    smilesSet.insert(smi);
+    if (!combinedMol.get()) {
+      combinedMol = boost::shared_ptr<RWMol>(new RWMol(*mols[0].get()));
+    } else {
+      ROMol *m = combineMols(*combinedMol.get(), *newMol.get());
+      combinedMol.reset(new RWMol(*m));
+      delete m;
+    }
+    smiles = getSmiles();
+  }
+
+  std::map<int,int> getNumBondsToRlabels() const {
+    std::map<int,int> rlabelsUsedCount;
+
+    for (ROMol::AtomIterator atIt = combinedMol->beginAtoms();
+         atIt != combinedMol->endAtoms();
+         ++atIt) {
+      Atom *atom = *atIt;
+      int rlabel; 
+      if(atom->getPropIfPresent<int>(RLABEL, rlabel))
+        rlabelsUsedCount[rlabel] += 1;
+    }
+    return rlabelsUsedCount;
+  }
+  
+  bool isHydrogen() const {  // is the rgroup all Hs
+    for(size_t i=0; i<mols.size(); ++i ) {
+      for (ROMol::AtomIterator atIt = mols[i]->beginAtoms();
+           atIt != mols[i]->endAtoms(); ++atIt) {
+        if ((*atIt)->getAtomicNum() > 1) return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  std::string getSmiles()
+      const {  // compute the canonical smiles for the attachments
+    std::string s;
+    for (std::set<std::string>::const_iterator it = smilesSet.begin();
+         it != smilesSet.end(); ++it) {
+      s += *it;
+    }
+    return s;
+  }
+};
+}
+
+namespace {
+struct RGroupMatch {
+  size_t core_idx;
+  std::map<int, RGroupData> rgroups;
+  RGroupMatch( size_t core_idx, const std::map<int, RGroupData> &rgroups ) :
+      core_idx(core_idx), rgroups(rgroups) {
+  }
+
+};
+
+struct CartesianProduct {
+  std::vector<size_t> permutation;
+  std::vector<size_t> sizes;
+  size_t maxPermutations;
+  size_t permutationCount;
+  CartesianProduct(const std::vector<size_t> &sizes)
+      : permutation(sizes.size(), 0), sizes(sizes), permutationCount(0) {
+    maxPermutations = 1;
+    for (size_t i = 0; i < sizes.size(); ++i)
+      maxPermutations *= sizes[i];  // may overflow....
+  }
+
+  bool next() {
+    ++permutationCount;
+    if (permutationCount == 1) {
+      return true;
+    }
+
+    return increment(0);
+  }
+
+  bool increment(size_t rowToIncrement) {
+    if (permutationCount > maxPermutations) return false;
+
+    permutation[rowToIncrement] += 1;
+    size_t max_index_of_row = sizes[rowToIncrement] - 1;
+    if (permutation[rowToIncrement] > max_index_of_row) {
+      permutation[rowToIncrement] = 0;
+      return increment(rowToIncrement + 1);
+    }
+    return true;
+  }
+};
+
+// stupid total score
```

Comment:
I do need to look at it again.  I kept tweaking it and now I probably don't understand what it does anymore.  In fact, you asking this question made me look into it again and I think there's a bug :)

What I *wanted* was essentially the following.  Let's assume that there are 10 mols.

```
R1 = 10 things the same should score 1.
R1 = (9 and 1) > (8 and 2) > (7 and 3) > (6 and 4) > (5 and 5). 
```

And I'm not sure I'm there yet actually and I think my "bug" is here:

```
    double tempScore = 1.;
    int maxMatches = 0;
    for (std::map<std::string, int>::const_iterator it = matchSet.begin();
         it != matchSet.end(); ++it) {
      if (it->second > maxMatches) { // XXXX BUG HERE
        tempScore *= it->second/(double)matches.size();
        maxMatches = it->second;
        m = it->first;
      }
```

See BUG HERE - this is a piddle from when I was just looking for the largest set.  Not combining all subsets.  Grrr.  This may be working by accident...